File: dotnet/Varar.Core/FailureAnchor.cs
namespace Varar.Core;

public static class FailureAnchor
{
    /// <summary>
    /// Where a failure points in the .md: a mismatch anchors at its first failing span (cell / doc
    /// string), anything else at the fallback (the step's match start). Port of <c>failure-anchor.ts</c>.
    /// </summary>
    public static Span Anchor(Exception? error, Span fallback) => error switch
    {
        CellMismatchError cm => cm.Cells.FirstOrDefault(c => !c.Ok)?.Span ?? fallback,
        _ => fallback,
    };

    /// <summary>
    /// The anchor travels with the thrown exception, from the executor (which knows the step) to
    /// <see cref="Failures.ToFailure"/> (which only sees the exception). TypeScript hangs it on the
    /// Error under a global symbol; here <see cref="Exception.Data"/> is the built-in equivalent, so
    /// no side table is needed.
    /// </summary>
    private const string AnchorKey = "varar.failureAnchor";

    /// <summary>Records on the exception itself where the failure points.</summary>
    public static void Attach(Exception? error, Span anchor)
    {
        // Data is read-only on a few framework exceptions; a failure that cannot carry its anchor
        // just falls back to the failing line, as it always did.
        if (error is null || error.Data.IsReadOnly)
        {
            return;
        }

        error.Data[AnchorKey] = anchor;
    }

    /// <summary>The anchor the executor attached, or <c>null</c> if there is none.</summary>
    public static Span? Attached(Exception? error) => error?.Data[AnchorKey] as Span;

    private const string DocPathKey = "varar.failureDocPath";

    /// <summary>
    /// Records the document the anchor's offsets belong to, for a step a reference block spliced in
    /// from another oath (ADR 0016). Travels the same way and for the same reason as the anchor:
    /// the executor knows the step, and whoever builds the failure payload sees only the error.
    /// </summary>
    public static void AttachDocPath(Exception? error, string docPath)
    {
        if (error is not null)
        {
            error.Data[DocPathKey] = docPath;
        }
    }

    public static string? AttachedDocPath(Exception? error) => error?.Data[DocPathKey] as string;
}

File: java/core/src/main/java/dev/varar/core/FailureAnchor.java
package dev.varar.core;

import java.util.Collections;
import java.util.Map;
import java.util.WeakHashMap;

/**
 * Where a failure POINTS in the {@code .md}: a mismatch anchors at its first failing span (the
 * cell, the doc string fence body), anything else at the fallback — the step's match start.
 *
 * <p>Port of {@code var-core/src/failure-anchor.ts}. This rule is the single source of truth
 * for failure locations: the executor's stack augmentation renders it per-runtime, and the
 * conformance trace pins it as {@code failure.anchor}, so every language port must reproduce
 * it byte-for-byte.
 */
final class FailureAnchor {
    private FailureAnchor() {}

    static Span anchor(Throwable error, Span fallback) {
        if (CellDiff.isCellMismatchException(error)) {
            for (CellDiff c : ((CellDiff.CellMismatchException) error).cells()) {
                if (!c.ok()) return c.span();
            }
            return fallback;
        }
        return fallback;
    }

    /**
     * The anchor has to travel with the thrown exception, from the executor (which knows the step)
     * to {@link Failure#toFailure} (which only sees the exception). TS hangs it on the Error under
     * a global symbol; a Java {@link Throwable} takes no extra fields, so this side table does the
     * same job. Weak keys, so remembering an anchor can never keep an exception alive;
     * synchronized, because a suite may run examples on several threads.
     */
    private static final Map<Throwable, Span> ATTACHED = Collections.synchronizedMap(new WeakHashMap<>());

    static void attach(Throwable error, Span anchor) {
        if (error != null) ATTACHED.put(error, anchor);
    }

    static Span attached(Throwable error) {
        return error == null ? null : ATTACHED.get(error);
    }
}
